RickAndMorty-iOS/Views/Characters/CharacterDetailsView.swift

```diff
@@ -0,0 +1,155 @@
+                                                      green: 0.75,
+                                                      blue: 0.66,
+                                                      alpha: 0.4)
+        cell.upperLabel.text = episode.name
+        cell.lowerLeftLabel.text = episode.episode
+        cell.lowerRightLabel.text = episode.air_date
+
+        for index in 0...3 {
+            let isIndexValid =  episode.characters.indices.contains(index)
+            if isIndexValid {
+                let urlString = episode.characters[index]?.image ?? ""
+                cell.characterAvatarImageViews[index].sd_setImage(with: URL(string: urlString))
+            }
+        }
+    }
+}
+
+// MARK: - CollectionView Layout
+extension CharacterDetailsView {
+    private func createLayout() -> UICollectionViewLayout {
+        let layout = UICollectionViewCompositionalLayout { (sectionIndex, _) -> NSCollectionLayoutSection? in
+
+            guard let sectionType = Section(rawValue: sectionIndex) else {
+                return nil
+            }
+
+            let columns = sectionType.columnCount
+
+            let itemSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0),
+                                                  heightDimension: .fractionalHeight(1.0))
+            let item = NSCollectionLayoutItem(layoutSize: itemSize)
+            item.contentInsets = NSDirectionalEdgeInsets(top: 5, leading: 5, bottom: 5, trailing: 5)
+
+            func getGroupHeight() -> NSCollectionLayoutDimension {
+                switch sectionType {
+                case .appearance, .empty:
+                    return NSCollectionLayoutDimension.estimated(280)
+                case .info, .location, .emptyInfo, .emptyLocation:
+                    return NSCollectionLayoutDimension.estimated(60)
+                default:
+                    return NSCollectionLayoutDimension.estimated(100)
+                }
+            }
+
+            let groupSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0),
+                                                   heightDimension: getGroupHeight())
+            let group = NSCollectionLayoutGroup.vertical(layoutSize: groupSize, repeatingSubitem: item, count: columns)
+
+            let section = NSCollectionLayoutSection(group: group)
+
+            let headerSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0), heightDimension: .estimated(50))
+            let header = NSCollectionLayoutBoundarySupplementaryItem(layoutSize: headerSize, elementKind: UICollectionView.elementKindSectionHeader, alignment: .top)
+            if self.collectionView.numberOfItems(inSection: sectionIndex) > 0 {
+                section.boundarySupplementaryItems = [header]
+            }
+            return section
+        }
+        return layout
+    }
+}
+
+extension CharacterDetailsView {
+    func titleView(image: String?, title: String!) -> UIView {
```

Comment:
why do we use '!' mark in parameter declaration?

Reply:
It is not safe to use it, because if you'll pass null here - there will be a crash.